File: src/Java/pipeline-core/src/main/java/com/pipeline/core/PipelineResult.java
package com.pipeline.core;

import java.util.List;
import java.util.Objects;

public record PipelineResult<C>(
    C context,
    boolean shortCircuited,
    List<PipelineError> errors,
    List<ActionTiming> actionTimings,
    long totalNanos
) {
  public PipelineResult {
    context = Objects.requireNonNull(context, "context");
    errors = List.copyOf(Objects.requireNonNull(errors, "errors"));
    actionTimings = List.copyOf(Objects.requireNonNull(actionTimings, "actionTimings"));
  }

  public boolean hasErrors() {
    return !errors.isEmpty();
  }
}

File: src/CSharp/pipeline_services/core/PipelineResult.cs
using System;
using System.Collections.Generic;

namespace PipelineServices.Core;

public sealed class PipelineResult<ContextType>
{
    public PipelineResult(
        ContextType contextValue,
        bool shortCircuited,
        IReadOnlyList<PipelineError> errors,
        IReadOnlyList<ActionTiming> actionTimings,
        long totalNanos)
    {
        Context = contextValue;
        ShortCircuited = shortCircuited;
        Errors = errors ?? throw new ArgumentNullException(nameof(errors));
        ActionTimings = actionTimings ?? throw new ArgumentNullException(nameof(actionTimings));
        TotalNanos = totalNanos;
    }

    public ContextType Context { get; }

    public bool ShortCircuited { get; }

    public IReadOnlyList<PipelineError> Errors { get; }

    public IReadOnlyList<ActionTiming> ActionTimings { get; }

    public long TotalNanos { get; }

    public bool HasErrors()
    {
        return Errors.Count > 0;
    }
}

